[traced-graph][sparse] propagate compressed sparsity

aartbik opened the pull request:
This PR proceeds with implementing the feature request https://github.com/pytorch/pytorch/issues/117188
by generalizing more cases that already work with COO to work with the compressed sparse formats as well.

cc @mrshenli @pritamdamania87 @zhaojuanmao @satgera @gqchen @aazzolini @osalpekar @jiayisuse @H-Huang @kwen2501 @awgu @fegin @XilunWu @wanchaol @fduwjj @wz337 @tianyu-l @wconstab @chauhang @d4l3k @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @peterbell10 @ipiszy @chenyang78 @kadeng @muchulee8 @ColinPeppler @amjames @desertfire

aartbik commented:
@pearu @ezyang 

This is ready for review now. It fixes a few missing cases for functional tensors on the compressed format for IdNet.
There is still an issue with to("meta") in the dynamo case, so it removes one skip but adds a smaller skip back.

Status for non-dynamo

Ran 151 tests
OK (skipped=80)

Status for dynamo:

Ran 151 tests 
OK (skipped=130)

So, overall, a step forward again! Please review this.

aartbik commented:
I also fixed the meta issue for dynamo, although I am not sure why this was not a problem for COO but is for the other compressed cases. In any case, now **no more skips** for IdNet, and we get:

Status both **with** and **without** dynamo:

Ran 151 tests
OK (skipped=80)

aartbik commented:
More compressed cases are handled now. The layout_impl() changes for the functional tensor wrapper makes sure the virtual method call actually goes to the right place (this was *documented* in the base class, but never correctly implemented). Adding sym_strides_custom() to the CSR implementation is perhaps the most controversial, but without this a lot of client cases will have to be cleaned up. Furthermore, COO does not forbid this method call, so it seems okay to keep this for the custom stride cases.

Current status, lots more test run, only 80 skip left to go:

Ran 152 tests
OK (skipped=80)

pearu commented on aten/src/ATen/SparseCsrTensorImpl.cpp:
```diff
@@ -239,7 +239,7 @@ IntArrayRef SparseCsrTensorImpl::strides_custom() const {
 }
 SymIntArrayRef SparseCsrTensorImpl::sym_strides_custom() const {
-  TORCH_CHECK(false, "Sparse ", at::sparse_csr::layoutToString(layout_, /*upper=*/true), " tensors do not have strides");
+  return TensorImpl::sym_strides_custom(); // TODO: remove override altogether?
```
No. Sparse tensors don't have strides (and storage_offset) by definition and returning the default strides here is wrong and meaningless. To but it differently, it is one of the methods to shoot yourself in the foot. See an alternative approach in my review summary.

aartbik replied:
I don't disagree, but... the COO case supports strides, since elem.stride() was returning (0,0,...) when layout was coo. Is that intentional then?

(not to argue, just for my own understanging ;-)

pearu replied:
> but... the COO case supports strides, since elem.stride() was returning (0,0,...)

IMHO, this is a historical mistake. The purpose of the strides-storage_offset pair is to compute the locations of tensor values in memory which is not applicable for sparse tensors (read: pytorch is not ready for [GCS](https://github.com/pearu/gcs/) that would make strides meaningful for sparse tensors as well).

pearu commented on aten/src/ATen/SparseCsrTensorImpl.cpp:
```diff
@@ -251,7 +251,7 @@ void SparseCsrTensorImpl::set_storage_offset(int64_t storage_offset) {
 }
 bool SparseCsrTensorImpl::is_contiguous_custom(MemoryFormat) const {
-  TORCH_CHECK(false, "Sparse ", at::sparse_csr::layoutToString(layout_, /*upper=*/true), " tensors do not have is_contiguous");
+  return false; // TODO: relay?
```
Same note about sparse strides applies here. Please restore the original check.

pearu reviewed (changes requested):
Overall looks good except we cannot define strides/storage_offset/contiguity for sparse tensors. The following patch provides an alternative approach: apply the proposition that strides/storage_offset is strided-tensor-only parameters:

<details>

```
diff --git a/torch/_functorch/_aot_autograd/functional_utils.py b/torch/_functorch/_aot_autograd/functional_utils.py
index 34624d7fc25..7615ff16281 100644
--- a/torch/_functorch/_aot_autograd/functional_utils.py
+++ b/torch/_functorch/_aot_autograd/functional_utils.py
@@ -311,8 +311,11 @@ def gen_alias_from_base(
 def has_same_metadata(t1, t2):
     return (
         definitely_true(sym_eq(t1.size(), t2.size()))
-        and definitely_true(sym_eq(t1.stride(), t2.stride()))
-        and definitely_true(t1.storage_offset() == t2.storage_offset())
+        and definitely_true(t1.layout == t2.layout)
+        and (is_sparse_any(t1) or
+             (definitely_true(sym_eq(t1.stride(), t2.stride()))
+              and definitely_true(t1.storage_offset() == t2.storage_offset()))
+            )
         and t1.is_conj() == t2.is_conj()
         and t1.is_neg() == t2.is_neg()
     )
diff --git a/torch/_subclasses/functional_tensor.py b/torch/_subclasses/functional_tensor.py
index 193c827a497..93769894cc9 100644
--- a/torch/_subclasses/functional_tensor.py
+++ b/torch/_subclasses/functional_tensor.py
@@ -14,6 +14,8 @@ from torch.utils._python_dispatch import (
     return_and_correct_aliasing,
     TorchDispatchMode,
 )
+from torch._subclasses.meta_utils import is_sparse_any
+
 
 not_implemented_log = torch._logging.getArtifactLogger(__name__, "not_implemented")
 
@@ -129,8 +131,8 @@ class FunctionalTensor(torch.Tensor):
             # We should probably eventually fix this so that the first overload can just handle dynamic shapes.
             cls,
             elem.shape,  # sizes
-            elem.stride(),  # strides
-            elem.storage_offset(),  # storage_offset
+            elem.stride() if not is_sparse_any(elem) else None,  # strides
+            elem.storage_offset() if not is_sparse_any(elem) else None,  # storage_offset
             None,  # memory_format
             elem.dtype,  # dtype
             elem.layout,  # layout
@@ -265,6 +267,10 @@ class FunctionalTensor(torch.Tensor):
     int = _conversion_method_template(dtype=torch.int32)
     long = _conversion_method_template(dtype=torch.int64)
 
+    @property
+    def layout(self):
+        return self.elem.layout
+
 
 class FunctionalTensorMode(TorchDispatchMode):
     def __init__(self, pre_dispatch=False, export=False, _allow_token_discovery=False):
diff --git a/torch/fx/experimental/proxy_tensor.py b/torch/fx/experimental/proxy_tensor.py
index cdd9995bbcd..03688fd4a47 100644
--- a/torch/fx/experimental/proxy_tensor.py
+++ b/torch/fx/experimental/proxy_tensor.py
@@ -29,6 +29,7 @@ from torch.utils._stats import count
 from torch.utils._traceback import CapturedTraceback
 import logging
 from torch._library.fake_class_registry import FakeScriptObject
+from torch._subclasses.meta_utils import is_sparse_any
 import warnings
 
 from torch.overrides import TorchFunctionMode
@@ -205,14 +206,16 @@ def track_tensor(tensor, proxy, *, constant, tracer):
         try_set_proxy_slot(s, lambda x, i: set_meta(
             tracer.create_proxy('call_function', torch.ops.aten.sym_size.int, (proxy, i), {}), x), i)
 
-    for i, s in enumerate(tensor.stride()):
-        try_set_proxy_slot(s, lambda x, i: set_meta(
-            tracer.create_proxy('call_function', torch.ops.aten.sym_stride.int, (proxy, i), {}), x), i)
+    if not is_sparse_any(tensor):
+        for i, s in enumerate(tensor.stride()):
+            try_set_proxy_slot(s, lambda x, i: set_meta(
+                tracer.create_proxy('call_function', torch.ops.aten.sym_stride.int, (proxy, i), {}), x), i)
 
     try_set_proxy_slot(tensor.numel(), lambda x: set_meta(
         tracer.create_proxy('call_function', torch.ops.aten.sym_numel.default, (proxy,), {}), x))
-    try_set_proxy_slot(tensor.storage_offset(), lambda x: set_meta(
-        tracer.create_proxy('call_function', torch.ops.aten.sym_storage_offset.default, (proxy,)), x))
+    if not is_sparse_any(tensor):
+        try_set_proxy_slot(tensor.storage_offset(), lambda x: set_meta(
+            tracer.create_proxy('call_function', torch.ops.aten.sym_storage_offset.default, (proxy,)), x))
     set_proxy_slot(tensor, tracer, _ProxyTensor(proxy, constant))
 
 def track_tensor_tree(inner_res, proxy_res, *, constant, tracer):
diff --git a/torch/fx/passes/shape_prop.py b/torch/fx/passes/shape_prop.py
index 3da0fdd76dc..dcaee3f8211 100644
--- a/torch/fx/passes/shape_prop.py
+++ b/torch/fx/passes/shape_prop.py
@@ -9,6 +9,7 @@ from torch.fx.node import Node, map_aggregate
 from typing import Any, Tuple, NamedTuple, Optional, Dict
 from torch.fx._compatibility import compatibility
 from torch._guards import detect_fake_mode
+from torch._subclasses.meta_utils import is_sparse_any
 
 __all__ = ['TensorMetadata', 'ShapeProp']
 
@@ -35,11 +36,11 @@ def _extract_tensor_metadata(result : torch.Tensor, include_contiguity=True) ->
     shape = result.shape
     dtype = result.dtype
     requires_grad = result.requires_grad
-    stride = result.stride()
+    stride = result.stride() if not is_sparse_any(result) else None
 
     memory_format = None
 
-    if include_contiguity:
+    if include_contiguity and not is_sparse_any(result):
         memory_formats = {
             torch.contiguous_format,
             torch.channels_last,

```

</details>

With the above patch,
```
pytest -sv test/export/test_sparse.py
```
passes OK with 72 passes and 80 skips.

aartbik commented:
PTAL

aartbik commented:
Hmm, this is a bit baffling. When executing the `self.assertEqual` after the to-meta, it call `SparseCsrTensorImpl::sym_strides_custom()` seemingly when executing this part of the code in the file `torch/testing/_comparison.py`:
```

   if (
        isinstance(actual, sequence_types)       <---
        and not isinstance(actual, str)
        and isinstance(expected, sequence_types)
        and not isinstance(expected, str)
    ):
```

Any idea how isinstance can actually try to access strides? When I add my hack back to allow strides, the crash goes away.

aartbik commented:
I am getting very strange corruptions with PYTORCH_TEST_WITH_DYNAMO=1, e.g. printing x before calling assertEqualMeta works fine, but printing this inside the method fails. Likewise, the code crashes in the middle of isinstance checks?!

Right before this happens, I am getting a lot of messages of the following form. Does that give a clue?

```
V0627 15:28:51.893000 140004138561600 torch/_dynamo/guards.py:2575] [15/2] [__recompiles] Recompiling function __instancecheck__ in <frozen abc>:117
V0627 15:28:51.893000 140004138561600 torch/_dynamo/guards.py:2575] [15/2] [__recompiles]     triggered by the following guard failure(s):
V0627 15:28:51.893000 140004138561600 torch/_dynamo/guards.py:2575] [15/2] [__recompiles]     - ___check_obj_id(L['cls'], 41642656)                         
V0627 15:28:51.893000 140004138561600 torch/_dynamo/guards.py:2575] [15/2] [__recompiles]     - expected type of 'L['instance']' to be a tensor type, ' but found <class 'torch._subclasses.fake_tensor.FakeTensor'>

```

aartbik commented:
I found it!  \o/

The dynamo guards were still accessing strides without any sparsity checks. When I added these, all tests pass, with and without dynamo runs:

```
$ PYTORCH_TEST_WITH_DYNAMO=0 python test/export/test_sparse.py

Ran 152 tests 
OK (skipped=80)

$ PYTORCH_TEST_WITH_DYNAMO=1 python test/export/test_sparse.py

Ran 152 tests 
OK (skipped=80)

```

PTAL @pearu @ezyang

pearu commented on torch/csrc/dynamo/guards.cpp:
```diff
@@ -157,20 +160,22 @@ std::string TensorCheck::check_verbose(
-      fail_reason << "stride mismatch at index " << i << ". expected "
-                  << known_stride.value() << ", actual " << strides[i];
-      return fail_reason.str();
+  if (!v.is_sparse() && at::sparse_csr::is_sparse_compressed(v)) {
```
Bug:
```suggestion
  if (!v.is_sparse() && !at::sparse_csr::is_sparse_compressed(v)) {
```

aartbik replied:
Yeah, good catch!

pearu commented on torch/_subclasses/meta_utils.py:
```diff
@@ -102,10 +102,13 @@ def go(m1, m2):
             assert_eq(m1.sparse_dim, m2.sparse_dim())
             assert_eq(m1.is_coalesced, m2.is_coalesced())
+        elif is_sparse_compressed(m1):
+            assert_eq(m1.sparse_dim, m2.sparse_dim())
```
Sparse compressed tensors have also dense dimensions:
```suggestion
            assert_eq(m1.sparse_dim, m2.sparse_dim())
            assert_eq(m1.dense_dim, m2.dense_dim())
```

aartbik replied:
added

pearu reviewed (changes requested):
There is a bug, otherwise looks good.

Thanks, @aartbik!

aartbik committed:
squash all commits

aartbik committed:
Merge branch 'pytorch:main' into bok

aartbik commented:
Apologies. My squash all commits seems to have pulled in "the world" for review ;-)
This is really just for @pearu and @ezyang to review

pearu commented on torch/_subclasses/fake_tensor.py:
```diff
@@ -1263,19 +1263,11 @@ def _prep_args_for_hash(self, args: Any) -> Any:
                     raise _BypassDispatchCache("constant attribute")
                 if arg.is_sparse:
                     raise _BypassDispatchCache("sparse tensor")
-                if arg.layout in [
-                    torch.sparse_csr,
-                    torch.sparse_csc,
-                    torch.sparse_bsr,
-                    torch.sparse_bsc,
-                ]:
-                    # Does this subsume arg.is_sparse?
-                    raise _BypassDispatchCache("sparse tensor layout")
+                if is_sparse_compressed(arg):
+                    raise _BypassDispatchCache("sparse compressed tensor")
                 # sparse tensors don't have storage, so check is after
                 if isinstance(arg.untyped_storage().nbytes(), torch.SymInt):
                     raise _BypassDispatchCache("symbolic nbytes")
-                if is_sparse_compressed(arg):
-                    raise _BypassDispatchCache("sparse compressed tensor")
```
Nit:
```suggestion
                if is_sparse_any(arg):
                    raise _BypassDispatchCache(f"{arg.layout} tensor")
                if isinstance(arg.untyped_storage().nbytes(), torch.SymInt):
                    raise _BypassDispatchCache("symbolic nbytes")

```

pearu reviewed (changes requested):
I have a minor nit.

For some reason, I cannot trigger CI tests. Could you make a small commit that hopefully resets something so that CI tests will be effective again.

aartbik committed:
collaps sparse exceptions into one branch

aartbik committed:
Merge branch 'pytorch:main' into bok

aartbik committed:
Merge branch 'pytorch:main' into bok

pearu commented:
It is not immediately clear why this PR affects mkldnn layout behavior. Could you try restoring aten/src/ATen/FunctionalTensorWrapper.cpp to see if mkldnn tests depend on it?

aartbik commented:
> It is not immediately clear why this PR affects mkldnn layout behavior. Could you try restoring aten/src/ATen/FunctionalTensorWrapper.cpp to see if mkldnn tests depend on it?

That change is still required to avoid the following (I was hoping you stride override made this change unnecessary, but alas):
```

  File "/usr/local/google/home/ajcbik/PYTORCH/pytorch/torch/_functorch/_aot_autograd/functional_utils.py", line 41, in to_fun
    return FunctionalTensor.to_functional(t)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/local/google/home/ajcbik/PYTORCH/pytorch/torch/_subclasses/functional_tensor.py", line 227, in to_functional
    out = FunctionalTensor(x_functional)
          ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/local/google/home/ajcbik/PYTORCH/pytorch/torch/_subclasses/functional_tensor.py", line 142, in __new__
    elem.stride() if not is_sparse_any(elem) else None,  # strides
                         ^^^^^^^^^^^^^^^^^^^
  File "/usr/local/google/home/ajcbik/PYTORCH/pytorch/torch/_subclasses/meta_utils.py", line 145, in is_sparse_any
    return is_sparse_coo(t) or is_sparse_compressed(t)
           ^^^^^^^^^^^^^^^^
  File "/usr/local/google/home/ajcbik/PYTORCH/pytorch/torch/_subclasses/meta_utils.py", line 128, in is_sparse_coo
    return isinstance(t, torch.Tensor) and t.layout is torch.sparse_coo
                                           ^^^^^^^^

```

pearu commented:
@aartbik , we'll still need to figure out why mkldnn tests fail with this PR. Can you confirm if restoring aten/src/ATen/FunctionalTensorWrapper.cpp will fix the mkldnn tests? Can you reproduce mkldnn test failures locally?

aartbik commented:
> @aartbik , we'll still need to figure out why mkldnn tests fail with this PR. Can you confirm if restoring aten/src/ATen/FunctionalTensorWrapper.cpp will fix the mkldnn tests? Can you reproduce mkldnn test failures locally?

on it!

aartbik committed:
Merge branch 'pytorch:main' into bok

aartbik committed:
Merge branch 'main' into bok

aartbik committed:
Merge branch 'pytorch:main' into bok

aartbik committed:
minor improvements

aartbik commented:
I made some minor improvements in the guards (not skipping sizes) and merged the layout_impl override with the mainline (no impact). If the problems still persist, I could some help with

(1) The failure in TestSparseMeta.test_add_meta_SparseBSC_float64. The layouts returned by x.layout, y.layout are different before and after calling the assertEqualMeta() check. I find this baffling, but have seen that before due to retriggered compilation?

(2) The mkldnn failure (if it is still there)..

@pearu terribly sorry to bug you with this (again), but do you have some clues on how to get this PR in?

aartbik committed:
reapply hot fix from mainline

aartbik committed:
Merge branch 'pytorch:main' into bok

aartbik committed:
Merge branch 'pytorch:main' into bok

amjames commented:
@aartbik Pearu is on holiday, so he might not be very responsive. I took a look at your failing tests and the mkldnn failure is popping up because:

1) FakeTensors created from mkldnn tensors don't properly populate the `layout` field (see [here](https://github.com/pytorch/pytorch/blob/042d764872a6307ed2f087cf8a741b2106370207/torch/_subclasses/meta_utils.py#L1165-L1171))
2) You have added a layout check to `assert_metadata_eq`. 

The real solution here would be for the layout field to be propagated to the FakeTensor correctly for `mkldnn`. However it appears that this is not supported. The combination of `meta` device type and `mkldnn` layout leads to an error. 

That is pretty far out of scope for this PR, so an alternative solution would be modify `assert_metadata_eq` to only check layout conditionally (ie. `is_any_sparse(m1) or is_any_sparse(m2)`). 

@ezyang wdyt?

aartbik committed:
disable test_add for dynamo (TBD on why it fails)

aartbik commented:
I added the mainline "hot-fix" back and added a dynamo-only skip for the add_meta failure. I am hopeful this passes everything now. Otherwise I will also do what you suggest above.

The good thing is that once this PR is in, we are only one minor PR away from getting ALL sparse export tests to pass and thus closing the feature request ;-)

aartbik committed:
Merge branch 'pytorch:main' into bok

aartbik committed:
only test meta layout for sparse (for now)

aartbik committed:
Merge branch 'pytorch:main' into bok

ezyang commented:
The test failures look legit

ezyang commented:
@pytorchbot rebase

aartbik commented:
Do you want me to squash all the commits into one to simplify rebase?

[I seem to be notorious bad at this, though, bringing in many other reviewers while the commits are stabilizing ;-)]

ezyang commented:
Just do a merge commit

aartbik committed:
Merge branch 'pytorch:main' into bok

aartbik committed:
Merge branch 'pytorch:main' into bok

aartbik committed:
fix "bitrot" that happened since last green run

aartbik committed:
more bitrot fix

aartbik committed:
Merge branch 'pytorch:main' into bok

aartbik committed:
Merge branch 'pytorch:main' into bok

aartbik committed:
guard newly added test

aartbik committed:
lint

aartbik committed:
Merge branch 'pytorch:main' into bok

aartbik committed:
lint

aartbik committed:
Merge branch 'pytorch:main' into bok

aartbik committed:
stricter "hot fix" back

aartbik commented:
This is frustrating. I feel that the new code is correct, but the removal of the hot fix in builder.py is what is causing all those new failures. Trying to put the old hot fix back with export guard to see if that helps....

We really need to get this code in to avoid the constant bit rotting about sparsity.

aartbik committed:
Merge branch 'pytorch:main' into bok

aartbik committed:
Merge branch 'pytorch:main' into bok

aartbik commented:
Just a heads up, I am going on a 10 day road trip and will have little to no access to a computer. So, although of course I would appreciate any help from core devs to get this PR submitted while I am absent, otherwise we will have to pick this up on July  22 when I am back.

aartbik committed:
Update builder.py

linting while on the road ;-)

pearu commented:
@aartbik I am OOO til July 28.

aartbik commented:
This has bitrotted too much. I am going to start a new branch and try again from scratch ;-(

aartbik commented:
I created a new PR https://github.com/pytorch/pytorch/pull/131474 that revolves all conflicts above, and also addresses some new issues due to new code being submitted along this export path. I am closing this PR in favor of the new shiny one....
